### ISSUE TITLE
AttributeError: 'WebSocketProtocol' object has no attribute 'reply_channel'

### ISSUE THREAD

**Issue by quinox** [andrewgodwin/daphne#28]
With an incorrect `nginx` setup (one that does not include `proxy_set_header Connection "upgrade";`) Daphne crashes resulting in a client-side **502 Bad gateway**.

When the log statement from `daphne/http_protocol.py:107` is removed Daphne no longer crashes resulting in a client-side **400 HTTP Connection headers do not include 'upgrade' value (case-insensitive) : close**, which is more helpful in determining what is wrong with the setup.

---

Stacktrace:

```python
2016-06-26 22:02:12,500 ERROR    Traceback (most recent call last):
  File "<...>/virtual/lib/python3.4/site-packages/daphne/http_protocol.py", line 107, in process
    logger.debug("Upgraded connection %s to WebSocket %s", self.reply_channel, protocol.reply_channel)
AttributeError: 'WebSocketProtocol' object has no attribute 'reply_channel'

Unhandled Error
Traceback (most recent call last):
  File "<...>/virtual/lib/python3.4/site-packages/twisted/python/log.py", line 101, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "<...>/virtual/lib/python3.4/site-packages/twisted/python/log.py", line 84, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "<...>/virtual/lib/python3.4/site-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "<...>/virtual/lib/python3.4/site-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "<...>/virtual/lib/python3.4/site-packages/twisted/internet/posixbase.py", line 597, in _doReadOrWrite
    why = selectable.doRead()
  File "<...>/virtual/lib/python3.4/site-packages/twisted/internet/tcp.py", line 209, in doRead
    return self._dataReceived(data)
  File "<...>/virtual/lib/python3.4/site-packages/twisted/internet/tcp.py", line 215, in _dataReceived
    rval = self.protocol.dataReceived(data)
  File "<...>/virtual/lib/python3.4/site-packages/twisted/protocols/basic.py", line 571, in dataReceived
    why = self.lineReceived(line)
  File "<...>/virtual/lib/python3.4/site-packages/twisted/web/http.py", line 1692, in lineReceived
    self.allContentReceived()
  File "<...>/virtual/lib/python3.4/site-packages/twisted/web/http.py", line 1781, in allContentReceived
    req.requestReceived(command, path, version)
  File "<...>/virtual/lib/python3.4/site-packages/twisted/web/http.py", line 768, in requestReceived
    self.process()
  File "<...>/virtual/lib/python3.4/site-packages/daphne/http_protocol.py", line 147, in process
    self.basic_error(500, b"Internal Server Error", "HTTP processing error")
  File "<...>/virtual/lib/python3.4/site-packages/daphne/http_protocol.py", line 256, in basic_error
    }).encode("utf8"),
  File "<...>/virtual/lib/python3.4/site-packages/daphne/http_protocol.py", line 216, in serverResponse
    http.Request.write(self, message['content'])
  File "<...>/virtual/lib/python3.4/site-packages/twisted/web/http.py", line 951, in write
    self.transport.writeSequence(l)
builtins.AttributeError: 'NoneType' object has no attribute 'writeSequence'
```

**Comment by andrewgodwin** [comment]
Could you provide a way to reproduce, preferably using `curl`? Usually leaving off the Upgrade header results in a `200 OK`, and the error you've posted a stack trace for looks like the `__init__` hasn't been run, which is... odd.

**Comment by quinox** [comment]
I'm happy to debug this further so let me know what I can do. To be explicit: This happens when `nginx` isn't configured with the Upgrade-header-settings; in both cases the `curl` command is identical and does contain the Upgrade-header.

---

I can reproduce it with this minimal setup:

**Terminal 1**
```bash
$ mkdir /tmp/channels
$ cd /tmp/channels
$ virtualenv --python /usr/bin/python3 virtual
$ . virtual/bin/activate
$ pip install django channels asgi_redis
$ django-admin startproject testing  
$ vim testing/testing/settings.py  
[add channels to INSTALLED_APPS + CHANNEL_LAYERS, SEE BOTTOM]

$ vim testing/asgi.py
[ADD asgi.py, SEE BOTTOM]

$ vim testing/testing/routing.py
[EMTPY ROUTES, SEE BOTTOM]

$ vim /etc/nginx/sites-enabled/mysite.conf  
[SEE BOTTOM]

$ cd testing  
$ PYTHONOPTIMIZE=1 daphne -p 16661 asgi:channel_layer  
```

**Terminal 2**

```bash
experience@reaper:~$ curl -i -N -H "Connection: Upgrade" -H "Upgrade: websocket" -H 'Sec-WebSocket-Version: 13'  -H 'Sec-WebSocket-Key: XXX==' -H "Host: experience.ytec.nl" -H "Origin: http://www.websocket.org" 'http://experience.ytec.nl/websocket/'
HTTP/1.1 502 Bad Gateway
Server: nginx/1.4.6 (Ubuntu)
Date: Mon, 27 Jun 2016 21:36:18 GMT
Content-Type: text/html
Content-Length: 181
Connection: keep-alive

<html>
<head><title>502 Bad Gateway</title></head>
<body bgcolor="white">
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx/1.4.6 (Ubuntu)</center>
</body>
</html>
```

(That is Daphne crashing)

**Terminal 1**

```bash
^C
$ vim /tmp/channels/virtual/lib/python3.4/site-packages/dhne/http_protocol.py +107
[comment log line out]
$ PYTHONOPTIMIZE=1 daphne -p 16661 asgi:channel_layer
```

**Terminal 2**

```bash
experience@reaper:~$ curl -i -N -H "Connection: Upgrade" -H "Upgrade: websocket" -H 'Sec-WebSocket-Version: 13'  -H 'Sec-WebSocket-Key: XXX==' -H "Host: experience.ytec.nl" -H "Origin: http://www.websocket.org" 'http://experience.ytec.nl/websocket/'
HTTP/1.1 400 HTTP Connection headers do not include 'upgrade' value (case-insensitive) : close
Server: nginx/1.4.6 (Ubuntu)
Date: Mon, 27 Jun 2016 21:37:33 GMT
Transfer-Encoding: chunked
Connection: keep-alive
```

---

`testing/testing/settings.py`:

```python
INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'channels',
]

CHANNEL_LAYERS = {
    'default': {
        'ROUTING': 'testing.routing.channel_routing',
        'BACKEND': 'asgi_redis.RedisChannelLayer',
        'CONFIG': {
            'hosts': [('localhost', 6379)],
        }
    }
}
```

--

`testing/asgi.py`:
```python
import os
from channels.asgi import get_channel_layer

os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'testing.settings')

assert not __debug__ #Make sure the server is configured for not running in debug mode. Just enable optimization to set __debug__ to false.

channel_layer = get_channel_layer()
```

---

`testing/routing.py`:
```python
channel_routing = []
```

---

`/etc/nginx/sites-enabled/mysite.conf`:
```config
server {
    listen 80;
    server_name experience.ytec.nl;

    location /websocket/ {
        proxy_pass http://localhost:16661;

        proxy_http_version 1.1;
        proxy_buffering off;  
        proxy_set_header upgrade $http_upgrade;
        #proxy_set_header Connection "upgrade";
    }
}
```

**Comment by andrewgodwin** [comment]
Thanks for the curl command - running it directly against daphne also helped reproduce the error, as it had a bad `Key` which went down the same path.

I've pushed a fix to master; if you could confirm it works, I'll make a release.

**Comment by quinox** [comment]
It works like a charm.

**Comment by alastairparagas** [comment]
I have the same problem as well. Tested the fix, seems good. If we can get it to release, that would be great.

**Comment by andrewgodwin** [comment]
Released in `0.13.1`.
